Display Avatar pictures in the Settingsdialog

dragotin opened the pull request:
This is a little beauty feature, to utilize owncloud/core#26872 . Hope you like it.

Looks like this:
![avatar](https://cloud.githubusercontent.com/assets/1070214/22182722/b57f9fec-e0ac-11e6-9f47-76deebb47368.png)

ogoffart commented on src/gui/settingsdialog.cpp:
```diff
@@ -236,6 +261,9 @@ void SettingsDialog::accountRemoved(AccountState *s)
         }
     }
 
+    if( _actionForAccount.contains(s->account().data()) ) {
```
Small detail:  No need of the if here.  remove does nothing if it is not there.

ogoffart commented on src/libsync/connectionvalidator.cpp:
```diff
@@ -252,7 +253,17 @@ void ConnectionValidator::slotUserFetched(const QVariantMap &json)
+
+        AvatarJob *job = new AvatarJob(_account, this);
+        QObject::connect(job, SIGNAL(avatarPixmap(QPixmap)), this, SLOT(slotAvatarPixmap(QPixmap)));
+
```
Missing  
```
    job->setTimeout(timeoutToUseMsec);
```

The default timeout is 5 minutes, and that's way to long for the connection validator which must finish much earlier in case of bad connection.


(And actually, since we increased the amount of jobs, timeoutToUseMsec might need to be lowered too.   But I'm wondering if it would be possible to fetch the capabilities, user name, and the avatar in parallel.)

dragotin replied:
I guess capabilities fetching could run in parallel to fetching the username, and following on that, the avatar, because the avatar needs the username. Maybe the fetch job of the avatar (at least) should be owned by the account, and could become fully async?

ogoffart commented on src/libsync/connectionvalidator.h:
```diff
@@ -119,6 +119,7 @@ protected slots:
 
     void slotCapabilitiesRecieved(const QVariantMap&);
     void slotUserFetched(const QVariantMap &);
+    void slotAvatarPixmap(const QPixmap&);
```
The ASCII art chart need to be updated at the beginning of this class.

dragotin commented:
Btw, I did not mean to push this to 2.3. Probably this PR should have gone to master, but your choice.

dragotin commented:
Updated the Ascii-art accordingly and set a 20 seconds timeout.

Also, I added code that cuts a circle out of the avatar image, similar to what the server does. Looks better.

ogoffart commented on src/libsync/networkjobs.cpp:
```diff
@@ -589,6 +590,42 @@ bool PropfindJob::finished()
+AvatarJob::AvatarJob(AccountPtr account, QObject *parent)
+    : AbstractNetworkJob(account, QString(), parent)
+{
+    _avatarUrl = Utility::concatUrlPath(account->url(), QString("remote.php/dav/avatars/%1/128.png").arg(account->davUser()));
```
Does not work for me with this URL. (owncloud master)
The URL that works for me is `index.php/avatar/%1/128`


@DeepDiver1975  What's the exact URL of the  avatars?

dragotin replied:
Hmm, I think the pull request that is implementing the avatar images is not yet merged in core: owncloud/core#26872

guruz replied:
Then let's delay this PR to master/2.4? sorry for that.

dragotin replied:
yes, sure.

ogoffart commented on src/libsync/connectionvalidator.cpp:
```diff
@@ -252,7 +253,18 @@ void ConnectionValidator::slotUserFetched(const QVariantMap &json)
         _account->setDavUser(user);
+
+        AvatarJob *job = new AvatarJob(_account, this);
+        job->setTimeout(20*1000);
```
(I think it should use the same timeout as all the other jobs in this file, for consistency)

dragotin replied:
@ogoffart wouldn''t it make more sense to set it in the constructor of AvatarJob anyway?

jturcotte commented on src/libsync/networkjobs.cpp:
```diff
@@ -589,6 +590,42 @@ bool PropfindJob::finished()
+{
+    int http_result_code = reply()->attribute(QNetworkRequest::HttpStatusCodeAttribute).toInt();
+
+    QPixmap avPixmap;
```
This seems to cause the FolderManTest to fail on your PR, since that needs a QGuiApplication. And you don't seem to need a QPixmap either, since you're creating a new one in `circleMask` anyway, so storing it as a QImage could be enough to fix that test.

dragotin commented:
5e33898a08b9b1b415562569793c9eef123a7d27 fixes the testsuite by using QImage instead of QPixmap

dragotin commented:
@ogoffart @jturcotte please consider merging

guruz commented:
(This would need to be merged in 2.4/master not 2.3)

guruz commented:
Bummer. Still needs all the ```SettingsDialogMac``` code.
@shadone ? :-P

guruz commented:
Why does Jenkins complain? https://jenkins.owncloud.org/job/owncloud-client/job/client/job/PR-5482/102/console

```
[ 13%] Building CXX object src/libsync/CMakeFiles/owncloudsync.dir/networkjobs.cpp.o
/var/lib/jenkins/workspace/cloud-client_client_PR-5482-YZILCXVUZD7LRF3ZRLYP2JS3Q6G7ZWAKK3GHZ7TY3RB7BIKKIFNA/src/libsync/networkjobs.cpp: In member function ‘virtual void OCC::AvatarJob::start()’:
/var/lib/jenkins/workspace/cloud-client_client_PR-5482-YZILCXVUZD7LRF3ZRLYP2JS3Q6G7ZWAKK3GHZ7TY3RB7BIKKIFNA/src/libsync/networkjobs.cpp:598:47: error: ‘davRequest’ was not declared in this scope
     setReply(davRequest("GET", _avatarUrl, req));
                                               ^
src/libsync/CMakeFiles/owncloudsync.dir/build.make:330: recipe for target 'src/libsync/CMakeFiles/owncloudsync.dir/networkjobs.cpp.o' failed
make[2]: *** [src/libsync/CMakeFiles/owncloudsync.dir/networkjobs.cpp.o] Error 1
CMakeFiles/Makefile2:1084: recipe for target 'src/libsync/CMakeFiles/owncloudsync.dir/all' failed
make[1]: *** [src/libsync/CMakeFiles/owncloudsync.dir/all] Error 2
Makefile:147: recipe for target 'all' failed
make: *** [all] Error 2
```

guruz commented:
It's not easy to adjust this for macOS since we also need API changes in src/3rdparty/qtmacgoodies :-|

I propose this gets merged first and then do a commit later for macOS support.

Code looks ok to me. So fine to merge in master/2.4

guruz commented:
I pushed the macOS support..

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/owncloud/client?pullRequest=5482) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/owncloud/client?pullRequest=5482) before we can accept your contribution.<br/>**2** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: dragotin<br/>:white_check_mark: guruz<br/>:x: Klaas Freitag<br/><hr/>**Klaas Freitag** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/owncloud/client?pullRequest=5482) it.</sub>

guruz commented on src/gui/settingsdialog.cpp:
```diff
@@ -196,8 +215,17 @@ void SettingsDialog::accountAdded(AccountState *s)
-    auto accountAction = createColorAwareAction(QLatin1String(":/client/resources/account.png"),
-                                                brandingSingleAccount ? tr("Account") : s->account()->displayName());
+    QAction *accountAction;
+    QImage avatar = s->account()->avatar();
```
I think this will never be called since accountAdded is called before the ConnectionValidator can run?

dragotin commented:
Hmm, I added a missing email address to my github account now again. Lets see if I have now signed the CLA again.

ogoffart commented:
I have the same error locally:

src/libsync/networkjobs.cpp:598:14: error: use of undeclared identifier 'davRequest'

ogoffart commented:
Merged in master via  commit	8cb3a77
